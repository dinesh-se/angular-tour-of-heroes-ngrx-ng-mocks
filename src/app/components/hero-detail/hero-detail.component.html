<div *ngIf="hero$ | async as hero">
  <h2>{{hero.name | uppercase}} Details</h2>
  <div><span>id: </span>{{hero.id}}</div>
  <div>
    <label for="hero-name">Hero name: </label>
    <input #heroName id="hero-name" placeholder="Hero name" value="{{hero.name}}" />
  </div>
  <button type="button" class="back-btn" (click)="goBack()">Go Back</button>
  <button type="button" class="save-btn" (click)="save(hero.id)">Save</button>
</div>
