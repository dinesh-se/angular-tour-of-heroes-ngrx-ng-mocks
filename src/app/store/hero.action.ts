import { createAction, props } from '@ngrx/store';

import { Hero } from '@model/hero';

export const loadHeroes = createAction('[Heroes] Load Heroes');

export const loadHeroesSuccess = createAction(
  '[Heroes] Load Heroes Success',
  props<{ heroes: Hero[] }>()
);

export const loadHeroesFailure = createAction(
  '[Heroes] Load Heroes Failure',
  props<{ error: any }>()
);

export const searchHeroes = createAction(
  '[Heroes] Search Heroes',
  props<{ query: string }>()
);

export const searchHeroesSuccess = createAction(
  '[Heroes] Search Heroes Success',
  props<{ searchResults: Hero[] }>()
);

export const searchHeroesFailure = createAction(
  '[Heroes] Search Heroes Failure',
  props<{ error: any }>()
);

export const addHero = createAction(
  '[Heroes] Add Hero',
  props<{ name: string }>()
);

export const addHeroSuccess = createAction(
  '[Heroes] Add Hero Success',
  props<{ hero: Hero }>()
);

export const addHeroFailure = createAction(
  '[Heroes] Add Hero Failure',
  props<{ error: any }>()
);

export const updateHero = createAction(
  '[Heroes] Update Hero',
  props<{ hero: Hero }>()
);

export const updateHeroSuccess = createAction(
  '[Heroes] Update Hero Success',
  props<{ hero: Hero }>()
);

export const updateHeroFailure = createAction(
  '[Heroes] Update Hero Failure',
  props<{ error: any }>()
);

export const deleteHero = createAction(
  '[Heroes] Delete Hero',
  props<{ id: number }>()
);

export const deleteHeroSuccess = createAction(
  '[Heroes] Delete Hero Success',
  props<{ id: number }>()
);

export const deleteHeroFailure = createAction(
  '[Heroes] Delete Hero Failure',
  props<{ error: any }>()
);

export const HeroActions = {
  loadHeroes,
  loadHeroesSuccess,
  loadHeroesFailure,
  searchHeroes,
  searchHeroesSuccess,
  searchHeroesFailure,
  addHero,
  addHeroSuccess,
  addHeroFailure,
  updateHero,
  updateHeroSuccess,
  updateHeroFailure,
  deleteHero,
  deleteHeroSuccess,
  deleteHeroFailure
};
