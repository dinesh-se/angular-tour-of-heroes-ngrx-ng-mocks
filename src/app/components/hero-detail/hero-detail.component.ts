import { Component, ElementRef, Input, OnInit, ViewChild } from '@angular/core';
import { AsyncPipe, NgIf, UpperCasePipe } from '@angular/common';
import { ActivatedRoute } from '@angular/router';
import { Location } from '@angular/common';
import { Store } from '@ngrx/store';
import { Observable } from 'rxjs';

import { Hero } from '@model/hero';
import { HeroesState } from '@store/hero.reducer';
import { HeroActions } from '@store/hero.action';
import { HeroSelectors } from '@store/hero.selector';

@Component({
  selector: 'app-hero-detail',
  standalone: true,
  imports: [UpperCasePipe, NgIf, AsyncPipe],
  templateUrl: './hero-detail.component.html',
  styleUrl: './hero-detail.component.scss'
})
export class HeroDetailComponent implements OnInit {
  hero$!: Observable<Hero | undefined>;

  @ViewChild('heroName') heroName: ElementRef | undefined;

  constructor(
    private route: ActivatedRoute,
    private location: Location,
    private store: Store<HeroesState>
  ) {}

  ngOnInit(): void {
    const id = Number(this.route.snapshot.paramMap.get('id'));
    this.hero$ = this.store.select(HeroSelectors.selectHeroById(+id));
  }

  goBack(): void {
    this.location.back();
  }

  save(id: number): void {
    const name = this.heroName?.nativeElement.value;

    if (name) {
      const hero: Hero = { id, name };
      this.store.dispatch(HeroActions.updateHero({ hero }));
      //TODO: refactor to go back when the state update is completed
      this.goBack();
    }
  }
}
