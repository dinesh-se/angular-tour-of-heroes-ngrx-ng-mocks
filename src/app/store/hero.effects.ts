import { Injectable } from '@angular/core';
import { Actions, createEffect, ofType } from '@ngrx/effects';
import { Observable, of } from 'rxjs';
import { catchError, map, switchMap, tap } from 'rxjs/operators';
import { Action } from '@ngrx/store';

import { HeroService } from '@services/hero.service';
import * as HeroActions from './hero.action';

@Injectable()
export class HeroEffects {
  constructor(
    private actions$: Actions,
    private heroService: HeroService
  ) {}

  loadHeroes$: Observable<Action> = createEffect(() =>
    this.actions$.pipe(
      ofType(HeroActions.loadHeroes),
      switchMap(() => this.heroService.getHeroes().pipe(
        tap((heroes) => console.log('HEROES', heroes)),
        map((heroes) => HeroActions.loadHeroesSuccess({ heroes })),
        catchError((error) => of(HeroActions.loadHeroesFailure({ error })))
      ))
    )
  );

  searchHeroes$: Observable<Action> = createEffect(() =>
    this.actions$.pipe(
      ofType(HeroActions.searchHeroes),
      switchMap(({ query }) =>
        this.heroService.searchHeroes(query).pipe(
          map((searchResults) => HeroActions.searchHeroesSuccess({ searchResults })),
          catchError((error) => of(HeroActions.searchHeroesFailure({ error })))
        )
      )
    )
  );

  addHero$: Observable<Action> = createEffect(() =>
    this.actions$.pipe(
      ofType(HeroActions.addHero),
      switchMap(({ name }) =>
        this.heroService.addHero({ name }).pipe(
          map((addedHero) => HeroActions.addHeroSuccess({ hero: addedHero })),
          catchError((error) => of(HeroActions.addHeroFailure({ error })))
        )
      )
    )
  );

  updateHero$: Observable<Action> = createEffect(() =>
    this.actions$.pipe(
      ofType(HeroActions.updateHero),
      switchMap(({ hero }) =>
        this.heroService.updateHero(hero).pipe(
          map(() => HeroActions.updateHeroSuccess({ hero })),
          catchError((error) => of(HeroActions.updateHeroFailure({ error })))
        )
      )
    )
  );

  deleteHero$: Observable<Action> = createEffect(() =>
    this.actions$.pipe(
      ofType(HeroActions.deleteHero),
      switchMap(({ id }) =>
        this.heroService.deleteHero(id).pipe(
          map(() => HeroActions.deleteHeroSuccess({ id })),
          catchError((error) => of(HeroActions.deleteHeroFailure({ error })))
        )
      )
    )
  );
}
