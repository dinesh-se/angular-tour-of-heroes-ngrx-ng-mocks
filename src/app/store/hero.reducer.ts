import { createReducer, on, Action } from '@ngrx/store';

import * as HeroActions from './hero.action';
import { Hero } from '@model/hero';

// Define the shape of the state
export interface HeroesState {
  searchResults: Hero[];
  heroes: Hero[];
  loading: boolean;
  error: any;
}

// Define the initial state
export const initialState: HeroesState = {
  searchResults: [],
  heroes: [],
  loading: false,
  error: null
};

// Create the reducer function using createReducer
const heroesReducer = createReducer(
  initialState,

  // Load Heroes actions
  on(HeroActions.loadHeroes, state => ({
    ...state,
    loading: true,
    error: null
  })),
  on(HeroActions.loadHeroesSuccess, (state, { heroes }) => ({
    ...state,
    heroes,
    loading: false
  })),
  on(HeroActions.loadHeroesFailure, (state, { error }) => ({
    ...state,
    loading: false,
    error
  })),


  on(HeroActions.searchHeroesSuccess, (state, { searchResults }) => ({
    ...state,
    searchResults,
  })),
  on(HeroActions.searchHeroesFailure, (state, { error }) => ({
    ...state,
    error
  })),

  // Add Hero actions
  on(HeroActions.addHeroSuccess, (state, { hero }) => ({
    ...state,
    heroes: [...state.heroes, hero]
  })),
  on(HeroActions.addHeroFailure, (state, { error }) => ({
    ...state,
    error
  })),

  // Update Hero actions
  on(HeroActions.updateHeroSuccess, (state, { hero }) => ({
    ...state,
    heroes: state.heroes.map(h => (h.id === hero.id ? hero : h))
  })),
  on(HeroActions.updateHeroFailure, (state, { error }) => ({
    ...state,
    error
  })),

  // Delete Hero actions
  on(HeroActions.deleteHeroSuccess, (state, { id }) => ({
    ...state,
    heroes: state.heroes.filter(h => h.id !== id)
  })),
  on(HeroActions.deleteHeroFailure, (state, { error }) => ({
    ...state,
    error
  }))
);

// Export the reducer function
export function reducer(state: HeroesState | undefined, action: Action) {
  return heroesReducer(state, action);
}
