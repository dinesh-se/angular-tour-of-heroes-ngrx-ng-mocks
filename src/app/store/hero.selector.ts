import { createSelector, createFeatureSelector } from '@ngrx/store';
import { HeroesState } from './hero.reducer';
import { Hero } from '@model/hero';

export const selectHeroesState = createFeatureSelector<HeroesState>('heroes');

export const selectAllHeroes = createSelector(
  selectHeroesState,
  (state: HeroesState) => state.heroes
);

export const selectHeroesLoading = createSelector(
  selectHeroesState,
  (state: HeroesState) => state.loading
);

export const selectHeroesError = createSelector(
  selectHeroesState,
  (state: HeroesState) => state.error
);

export const selectSearchResults = createSelector(
  selectHeroesState,
  (state: HeroesState) => state.searchResults
);

export const selectQuery = createSelector(
  selectHeroesState,
  (state: HeroesState) =>  state.query
);

export const selectHeroById = (id: number) => createSelector(
  selectAllHeroes,
  (heroes: Hero[]) => heroes.find(hero => hero.id === id)
);
