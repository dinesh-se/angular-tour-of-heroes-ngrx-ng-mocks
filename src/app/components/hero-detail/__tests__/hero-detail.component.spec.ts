import { AsyncPipe, NgIf, UpperCasePipe } from '@angular/common';
import { ActivatedRoute, convertToParamMap } from '@angular/router';
import { Store } from '@ngrx/store';
import { MockStore, createMockStore } from '@ngrx/store/testing';
import { MockBuilder, MockRender, MockedComponentFixture } from 'ng-mocks';
import { of } from 'rxjs';

import { HeroesState } from '@store/hero.reducer';
import { HeroSelectors } from '@store/hero.selector';
import { HeroActions } from '@store/hero.action';
import { HeroDetailComponent } from '../hero-detail.component';

describe('HeroDetailComponent', () => {
  let component: HeroDetailComponent;
  let fixture: MockedComponentFixture<HeroDetailComponent>;
  const store: MockStore<HeroesState> = createMockStore<HeroesState>();
  const selectHeroByIdSpy = jest.spyOn(HeroSelectors, 'selectHeroById');
  const dispatchSpy = jest.spyOn(store, 'dispatch');
  const updateHeroSpy = jest.spyOn(HeroActions, 'updateHero');
  const selectHeroSpy = jest.spyOn(store, 'select')
    .mockImplementation(() => {
      return of({
        id: '1',
        name: 'Hero 1'
      });
    });

  beforeEach(async () => {
    await MockBuilder(HeroDetailComponent)
      .provide({
        provide: ActivatedRoute,
        useValue: {
          snapshot: {
            paramMap: convertToParamMap({ id: 1 })
          }
        }
      })
      .mock(Store, store, { precise: true })
      .keep(NgIf)
      .keep(AsyncPipe)
      .keep(UpperCasePipe);

    fixture = MockRender(HeroDetailComponent);
    component = fixture.point.componentInstance;
    fixture.detectChanges();
  });

  afterEach(() => {
    jest.clearAllMocks();
  });

  it('should create with hero details', () => {
    const input = fixture.point.nativeElement.querySelector('input');

    expect(component).toBeTruthy();
    expect(selectHeroSpy).toHaveBeenCalled();
    expect(selectHeroByIdSpy).toHaveBeenCalledWith(1);
    expect(input.value).toBe('Hero 1');
    expect(fixture).toMatchSnapshot();
  });

  it('should save updated hero details', () => {
    const goBackSpy = jest.spyOn(component, 'goBack');
    const saveButton = fixture.point.nativeElement.querySelector('.save-btn');
    const input = fixture.point.nativeElement.querySelector('input');
    input.value = 'Updated Hero 1';
    const updatedHero = {
      hero: {
        id: "1",
        name: 'Updated Hero 1'
      }
    };

    saveButton.click();

    expect(dispatchSpy).toHaveBeenCalled();
    expect(updateHeroSpy).toHaveBeenCalledWith(updatedHero);
    expect(goBackSpy).toHaveBeenCalled();
  });

  it('should not save hero details when hero name is empty', () => {
    const goBackSpy = jest.spyOn(component, 'goBack');
    const saveButton = fixture.point.nativeElement.querySelector('.save-btn');
    const input = fixture.point.nativeElement.querySelector('input');
    input.value = '';

    saveButton.click();

    expect(dispatchSpy).not.toHaveBeenCalled();
    expect(updateHeroSpy).not.toHaveBeenCalled();
    expect(goBackSpy).not.toHaveBeenCalled();
  });
});
