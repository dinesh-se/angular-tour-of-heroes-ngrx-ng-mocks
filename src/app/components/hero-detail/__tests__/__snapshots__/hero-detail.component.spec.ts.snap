// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`HeroDetailComponent should create with hero details 1`] = `
<mock-render>
  <app-hero-detail>
    <div>
      <h2>
        HERO 1 Details
      </h2>
      <div>
        <span>
          id: 
        </span>
        1
      </div>
      <div>
        <label
          for="hero-name"
        >
          Hero name: 
        </label>
        <input
          id="hero-name"
          placeholder="Hero name"
        />
      </div>
      <button
        class="back-btn"
        type="button"
      >
        Go Back
      </button>
      <button
        class="save-btn"
        type="button"
      >
        Save
      </button>
    </div>
  </app-hero-detail>
</mock-render>
`;
